readme.md
# Botkit CMS

## An open tool for designing, building and managing interactive dialog systems

This project can serve as a drop-in replacement for users of Botkit Studio.

![Dialog Editor](https://raw.githubusercontent.com/howdyai/botkit-cms/master/docs/screenshots/dialog.png)

## Create Botkit Dialog Editor & API Service

Clone this repo and set it up on a public host somewhere. Clicking the Glitch link below will do this for you.

[Configure the .env file.](#configuration)

If migrating from Botkit Studio, [get your existing scripts](#migrate-from-botkit-studio), and put the resulting `scripts.json` file in the `.data/` folder or import it through the UI.

Launch the app, then load it in your web browser. You should see a link to the editor screen.

Make sure your new service is available at a public address on the web. Then, modify your Botkit app to include a pointer to this new service.

```
var controller = Botkit.platform({
  studio_command_uri: 'https://my.new.service',
  studio_token: 'a shared secret'
})
```

[![Remix on Glitch](https://cdn.glitch.com/2703baf2-b643-4da7-ab91-7ee2a2d00b5b%2Fremix-button.svg)](https://glitch.com/edit/#!/remix/botkit-cms)

## Migrate from Botkit Studio

If you have existing script content in Botkit Studio, you can migrate to a Botkit CMS instance by using the data export tool.

* Login to your Botkit Studio account.  
* Navigate to a specific bot and find the "Migrate Now" button on the dashboard.
* Deploy the service to Glitch - this will preconfigure the service with your current `studio_token` and other settings.
* Click "Download" to download the file, or copy paste the content of the window into a local file named `scripts.json`
* Use the "Import" tool within Botkit CMS to import the scripts.

### Configuration

Options for this service are controlled via environment variables, which can be stored in a `.env` file at the project root.

Here is an example `.env` file:

```
# Chat platform
# PLATFORM=[slack,teams,ciscospark,web,facebook]
PLATFORM=slack

# authentication tokens for Bot users
# TOKENS="123 456"
TOKENS=youwillneverguessmysecretbottoken

# Admin users for UI
# USERS="username:password username2:password2 username3:password3"
USERS=admin:123secret

# LUIS Endpoint
# URL to published LUIS Endpoint in the form of https://<REGION>.api.cognitive.microsoft.com/luis/v2.0/apps/<APPID>?subscription-key=<SUBID>&verbose=true&timezoneOffset=-360&q=
# Get this from LUIS.ai Keys and Endpoint settings
# LUIS_ENDPOINT=

# LUIS App Version
# Defaults to 0.1, update if you manage multiple LUIS app versions
# LUIS_VERSION=0.1
```

### Using LUIS.ai

This project includes support for using LUIS.ai to determine the intent represented by an incoming message.
To enable LUIS, add the `LUIS_ENDPOINT` variable to your environment.

After enabling LUIS, new options will appear in the Trigger dialog that will allow you to assign intents from LUIS as triggers.

### Editor Configuration

The Botkit dialog editor can be used in one of several different flavors, controlled by the `PLATFORM` environment variable.

| Value | Description
|--- |---
| web | Botkit Anywhere mode
| slack | Slack mode
| teams | Microsoft Teams mode
| ciscospark | Cisco Spark / Webex Teams mode
| facebook | Facebook mode


### Securing Admin / Editor Access

Access can be limited and users can be controlled using the `USERS` environment variable.

Set the variable to a space separated list of user:password pairs. Users will be shown a login prompt when accessing any url within the `/admin/` url.

### Securing API Access

You can lock down access to the API by specifying one or more access tokens in the TOKENS environment variable (or in the .env file).  

If any tokens are specified, access to the API requies a valid value in the `access_token` url query parameter.  Botkit will automatically use the Botkit Studio `studio_token` value for this.

## Building Project

Modifications to the front end application or css should be done to their original source files, then compiled by the build process. To build the Javascript and CSS files from their source locations, run the following command:

```bash
npm run build
```

The front end editor application included in this project is built with Angular. The source code of the this application is broken up into several component files in the `js/` folder. These are compiled into a single source file and moved to a final home at `public/js/scripts.js`  by the build process.

The CSS is controlled by SASS files in the `sass/` folder. These are compiled into a single source file and moved to a final home at `public/css/new.css`  by the build process.


## Alternate: Use this as a component in your Botkit App

First, npm install this:

```bash
npm install --save botkit-cms
```

[Get your existing scripts](#get-script-content), and put the resulting `scripts.json` in the main folder of your bot project.

Then, add to your bot's main file, just after defining your controller object:

```js
var cms = require('botkit-cms')();
cms.useLocalStudio(controller);

cms.loadScriptsFromFile(__dirname + '/scripts.json').catch(function(err) {
  console.error('Error loading scripts', err);
});
```


CHANGELOG.md
# 0.0.3

* Bug fixes
* Changed location of the sample_scripts.json file
* Clarified some docs

# 0.0.2 

First public release

# 0.0.1

Initial version

components/routes/admin.js

const uuidv4 = require('uuid/v4');
const request = require('request');
const async = require('async');

module.exports = function(webserver, api) {

    webserver.get('/admin', function(req, res) {
        res.render('index',{
            layout: 'layouts/layout',
        });
    });

    webserver.get('/admin/edit/:name', function(req, res) {
        res.render('edit',{
            layout: 'layouts/layout',
            platform: process.env.PLATFORM || 'web',
            command_id: req.params.name,
        });
    });

    webserver.post('/admin/save', function(req, res) {
        var update = req.body;

        api.getScripts().then(function(scripts) {
            var found = false;
            for (var s = 0; s < scripts.length; s++) {
                if (scripts[s].id === update.id) {
                    found = s;
                }
            }
            
            if (update.is_fallback) {
                scripts.forEach(function(script) {
                    script.is_fallback = false;
                });
            }

            if (found === false) {
                if (!update.id && update._id) {
                    update.id = update._id;
                } else if (!update.id) {
                    update.id = uuidv4();
                }
                update.modified = new Date();
                scripts.push(update);
                api.writeScriptsToFile(scripts).then(function() {
                    res.json({
                        success: true,
                        data: update,
                    });
                });

            } else if (new Date(scripts[found].modified) > new Date(update.modified)) {

                // if the version in the database was more recently modified, reject this update!
                res.json({
                    success: false,
                    message: 'Script was modified more recently, please refresh your browser to load the latest',
                });

            } else {

                var original_name = scripts[found].command;

                scripts[found] = update;
                scripts[found].modified = new Date();

                if (update.command != original_name) {
                    handleRenamed(scripts, original_name, update.command).then(function() {
                        api.writeScriptsToFile(scripts).then(function() {
                            res.json({
                                success: true,
                                data: update,
                            });
                        });
                    });
                } else {
                    api.writeScriptsToFile(scripts).then(function() {
                        res.json({
                            success: true,
                            data: update,
                        });
                    });
                }
            }
        });

    });


    function handleRenamed(scripts, original_name, new_name) {
        return new Promise(function(resolve, reject) {
            async.each(scripts, function(command, next) {
                updateExecScript(command, original_name, new_name, next);
            }, function() {
                resolve();
            })
        });
    }

    function updateExecScript(command, original_name, new_name, next) {
        // need to look at command.script[*].script[*].action
        // need to look at command.script[*].script[*].collect.options[*].action
        var dirty = false;
        for (var t = 0; t < command.script.length; t++) {
            for (var m = 0; m < command.script[t].script.length; m++) {
                if (command.script[t].script[m].action == 'execute_script' && command.script[t].script[m].execute && command.script[t].script[m].execute.script == original_name) {
                    command.script[t].script[m].execute.script = new_name;
                    dirty = true;
                }

                if (command.script[t].script[m].collect && command.script[t].script[m].collect.options) {
                    for (var o = 0; o < command.script[t].script[m].collect.options.length; o++) {
                        if (command.script[t].script[m].collect.options[o].action=='execute_script' && command.script[t].script[m].collect.options[o].execute && command.script[t].script[m].collect.options[o].execute.script == original_name) {
                            command.script[t].script[m].collect.options[o].execute.script = new_name;
                            dirty = true;
                        }
                    }
                }
            }
        }

        if (dirty) {
            command.modified = new Date();
            next();
        } else {
            next();
        }
    }



    // receives: command, user
    webserver.post('/admin/api/script', function(req, res) {
        if (req.body.command) {
            api.getScript(req.body.command).then(function(script) {
                res.json({success: 'ok', data: script});
            }).catch(function(err) {
                if (err) {
                    console.error('Error in getScript',err);
                }
                res.json({});
            })
        } else if (req.body.id) {
            api.getScriptById(req.body.id).then(function(script) {
                res.json(script);
            }).catch(function(err) {
                if (err) {
                    console.error('Error in getScript',err);
                }
                res.json({});
            })
        }
    });


    // receives: command, user
    webserver.get('/admin/api/scripts', function(req, res) {
        api.getScripts(req.query.tag).then(function(scripts) {
            res.json({success: true, data: scripts});
        }).catch(function(err) {
            if (err) {
                console.error('Error in getScripts',err);
            }
            res.json({});
        })
    });

    function parseLUISEndpoint(endpoint) {

        // endpoint in form of 
        // https://westus.api.cognitive.microsoft.com/luis/v2.0/apps/<APPID>?subscription-key=<SUBID>&verbose=true&timezoneOffset=-360&q=

        const config = {
            region: 'westus',
            app: '',
            key: '',
        }

        config.region = endpoint.match(/https\:\/\/(.*?)\.api/)[1];
        config.app = endpoint.match(/\/apps\/(.*?)\?/)[1];
        config.key = endpoint.match(/subscription\-key\=(.*?)\&/)[1];
        config.version = process.env.LUIS_VERSION || "0.1";

        return config;

    }

    webserver.get('/admin/api/luisIntents', function(req, res) {
        if (process.env.LUIS_ENDPOINT) {

            const luisConfig = parseLUISEndpoint(process.env.LUIS_ENDPOINT);
            var url = `https://${ luisConfig.region }.api.cognitive.microsoft.com/luis/api/v2.0/apps/${ luisConfig.app }/versions/${ luisConfig.version }/intents?skip=0&take=500`;

            var options = {
                method: 'GET',
                url: url,
                headers: {
                    'Ocp-Apim-Subscription-Key': luisConfig.key
                }
            };
            request(options, function(err, resp, body) {
                if (err) {
                    console.error('Error commnicating with LUIS:', err);
                    res.json({success: true, data: []});
                } else {
                    var intents = [];
                    try {
                        intents = JSON.parse(body);
                    } catch(err) {
                        console.error('Error parsing LUIS intents:', err);
                    }
                    res.json({success: true, data: intents});
                }                  
            });
          
        } else {
            res.json({success: true, data: []});
        }

    });

    webserver.delete('/admin/api/scripts/:id', function(req, res) {
        api.getScripts().then(function(scripts) {

            // delete script out of list.
            scripts = scripts.filter((script) => { return (script.id !== req.params.id) });

            // write scripts back to file.
            api.writeScriptsToFile(scripts).then(function() {
                res.json({
                    success: true,
                    data: scripts,
                });
            });

        }).catch(function(err) {
            if (err) {
                console.error('Error in getScripts',err);
            }
            res.json({});
        })
    });

    webserver.get('/admin/config', function(req, res) {

        var allowed_tokens = process.env.TOKENS ? process.env.TOKENS.split(/\s+/) : [];

        var package_version = require('../../package.json').version;

        res.render('config',{
            layout: 'layouts/layout',
            token: allowed_tokens[0],
            tokens: allowed_tokens.join("\n"),
            token_count: allowed_tokens.length,
            multiple_tokens: (allowed_tokens.length > 1),
            version: package_version,
            url: req.protocol + "://" + req.headers.host
        });
    })


}
